web/src/submessage.ts

```diff
@@ -1,9 +1,57 @@
+import z from "zod";
+
 import * as blueslip from "./blueslip";
 import * as channel from "./channel";
 import * as message_store from "./message_store";
+import type {Message} from "./message_store";
+import type {Submessage} from "./types";
 import * as widgetize from "./widgetize";
 
-export function get_message_events(message) {
+export const zform_widget_extra_data_schema = z.object({
+    choices: z.array(
+        z.object({
+            type: z.string(),
+            long_name: z.string(),
+            reply: z.string(),
+            short_name: z.string(),
+        }),
+    ),
+    heading: z.string(),
+    type: z.literal("choices"),
+});
+
+const poll_widget_extra_data_schema = z.object({
+    question: z.string().optional(),
+    options: z.array(z.string()).optional(),
+});
```

Comment:
Any specific reason for keeping both fields here optional? I need to add a similar extra data schema for the todo widget in #24235 and if I keep the fields there optional too, the union of these together does not pick up the correct schema since the 1st of the 2 basically matches with both. This also limits adding more extra data types in the future.

Would appreciate a speedy response :)

Reply:
I referred to #27317 and the already created `ExtraData` in `poll_widget.ts` (since it was already [migrated](https://github.com/zulip/zulip/blob/8a807949df4ef44d8d403a09815c2ca20a622c50/web/src/poll_widget.ts#L21)) for creating this schema.

Reply:
But looking back, I think these fields **can't** be actually `.optional( or undefined)`. When the user creates a `poll` but does not put a question or an option, in that case `question` is going to be empty `string` and `options` is going to be an `empty array`.

Reply:
After playing with `/poll` widget, I think we should remove those `.optional`s . What's your opinion on this (since you know about this subsystem **a lot** more that I do) ?

Reply:
> On some more thinking I realise that these field will need to be optional for backwards compatibility, to continue to support old polls the extra data for which did not have these field.

Ok. Makes sense :)

> Since I'm not too familiar with Zod, I'm not sure but I think [discriminated union](https://zod.dev/?id=discriminated-unions) can be used instead here to apply the correct extra data type based on the widget_type. What do you think?

> Edit: Simply making data a union of each widget_type's data instead of making extra_data a union might work even better and cleaner?

Yeah. And `widget_type` can act as a _discriminator key_ (which will help TypeScript infer the exact widget type (poll, todo etc.) and the corresponding `extra_data` object)

Reply:
Something like this ?
```
const data=z.discriminatedUnion("widget_type",[poll_widget_data, zform_data, foo_data..])


```

Reply:
@N-Shar-ma I suppose you are working on the above discussed refactor OR do you want me to raise a PR fixing this. I will be happy to provide any assistance :smile:

Reply:
Yes I'll work on a refactor based on this discussion, and tag you for review once I've got it to work. Thanks!